tests/integration/security/reachability/reachability_test.go

```diff
@@ -44,280 +43,192 @@ import (
-							},
-							ExpectSuccess: true,
-						},
+					include: func(src echo.Instance, opts echo.CallOptions) bool {
+						// Include all tests that target app B, which has the single-port config.
+						return opts.Target == b
+					},
+					expectSuccess: func(src echo.Instance, opts echo.CallOptions) bool {
+						return opts.PortName != "http"
 					},
 				},
 			}
 
 			for _, c := range testCases {
 				testName := strings.TrimSuffix(c.configFile, filepath.Ext(c.configFile))
-				t.Run(testName, func(t *testing.T) {
+				ctx.NewSubTest(testName).
+					Run(func(ctx framework.TestContext) {
+						// Apply the policy.
+						policyYAML := file.AsStringOrFail(ctx, filepath.Join("testdata", c.configFile))
+						g.ApplyConfigOrFail(ctx, c.namespace, policyYAML)
+						ctx.WhenDone(func() error {
+							return g.DeleteConfig(ctx, c.namespace.Name(), policyYAML)
+						})
 
-					// Apply the policy.
-					deploymentYAML := file.AsStringOrFail(t, filepath.Join("testdata", c.configFile))
-					g.ApplyConfigOrFail(t, c.namespace, deploymentYAML)
-					defer g.DeleteConfigOrFail(t, c.namespace, deploymentYAML)
+						// Give some time for the policy propagate.
+						// TODO: query pilot or app to know instead of sleep.
+						time.Sleep(10 * time.Second)
 
-					// Give some time for the policy propagate.
-					// TODO: query pilot or app to know instead of sleep.
-					time.Sleep(time.Second)
+						for _, src := range []echo.Instance{a, b, headless, naked} {
+							for _, dest := range []echo.Instance{a, b, headless, naked} {
+								for _, opts := range callOptions {
+									// Copy the loop variables so they won't change for the subtests.
+									src := src
+									dest := dest
+									opts := opts
 
-					for _, subTest := range c.subTests {
-						subTestName := fmt.Sprintf("%s->%s:%s",
-							subTest.From.Config().Service,
-							subTest.Options.Target.Config().Service,
-							subTest.Options.PortName)
-						t.Run(subTestName, func(t *testing.T) {
-							retry.UntilSuccessOrFail(t, subTest.Check, retry.Delay(time.Second), retry.Timeout(10*time.Second))
-						})
-					}
-				})
+									// Set the target on the call options.
+									opts.Target = dest
+
+									if c.include(src, opts) {
+										expectSuccess := c.expectSuccess(src, opts)
+
+										subTestName := fmt.Sprintf("%s->%s://%s:%s",
+											src.Config().Service,
+											opts.Scheme,
+											dest.Config().Service,
+											opts.PortName)
+
+										ctx.NewSubTest(subTestName).
+											RunParallel(func(ctx framework.TestContext) {
+												checker := connection.Checker{
+													From:          src,
+													Options:       opts,
+													ExpectSuccess: expectSuccess,
+												}
+												checker.CheckOrFail(ctx)
+											})
+									}
+								}
+							}
+						}
+					})
 			}
 		})
 }
+
+func config(name string, ns namespace.Instance, headless bool, annos echo.Annotations) echo.Config {
```

Comment:
Should we make this function for general to make it a utility function? Looks like a lot of places have similar code for this.

Reply:
Also rename it to make it more descriptive.

Reply:
we can consider doing that later on after we have multiple call sites.

Reply:
It's probably not entirely necessary, but it preserves the tests from the old pilot e2e tests.

Reply:
AFAIK, there are multiple callers. I can refactor this later :)